File: packages/node-sdk/src/runtime.ts
import { spawn, ChildProcess, execFile } from "node:child_process";
import { access, mkdir, unlink, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { homedir } from "node:os";
import { ProxyAgent } from "undici";
import { NeonClient } from "./client.js";

export const NEON3_RUNTIME_VERSION = "latest";
const NEON3_RUNTIME_REPOSITORY = "unco999/Neon3-CiJian";
const NEON3_RUNTIME_ASSET = (version: string) => `neon3-runtime-windows-x86_64-${version}.zip`;

export type RuntimeMode = "windowed" | "headless" | "external_surface";
export type RuntimeProfile = "auto" | "debug" | "release";
export interface RuntimeConfig { neonRoot?: string; mode?: RuntimeMode; profile?: RuntimeProfile; runtimeVersion?: string; eventd?: string; ui?: string; wgpu?: string; domain?: string; timeoutMs?: number; }

export class RuntimeSession {
  private processes: ChildProcess[] = [];
  readonly config: Required<RuntimeConfig>;
  private executableDir = "";
  private selectedProfile: Exclude<RuntimeProfile, "auto"> = "debug";
  constructor(config: RuntimeConfig) {
    this.config = { neonRoot: defaultNeonRoot(), mode: "windowed", profile: (process.env.NEON_PROFILE as RuntimeProfile | undefined) ?? "auto", runtimeVersion: process.env.NEON3_RUNTIME_VERSION ?? NEON3_RUNTIME_VERSION, eventd: "127.0.0.1:39101", ui: "127.0.0.1:39102", wgpu: "127.0.0.1:39103", domain: "127.0.0.1:39104", timeoutMs: 15000, ...config };
  }
  get activeProfile(): Exclude<RuntimeProfile, "auto"> { return this.selectedProfile; }
  async start(): Promise<void> {
    if (this.config.runtimeVersion === "latest") {
      const version = await resolveLatestRuntimeVersion();
      this.config.neonRoot = defaultNeonRoot(version);
      this.config.runtimeVersion = version;
    }
    this.executableDir = await this.findExecutableDir();
    const specs: Array<[string, string, string[]]> = [
      ["eventd", join(this.executableDir, "neon-eventd.exe"), ["--server", this.config.eventd, "1"]],
      ["wgpu", join(this.executableDir, "neon-wgpu-runtime.exe"), this.wgpuArgs()],
      ["ui", join(this.executableDir, "neon-ui-runtime.exe"), ["--forward-server", this.config.ui, this.config.wgpu, this.config.domain, "--eventd", this.config.eventd]],
    ];
    try {
      for (const [name, executable, args] of specs) { await access(executable); this.processes.push(spawn(executable, args, { cwd: this.config.neonRoot, stdio: "ignore", windowsHide: false })); void name; }
      await this.waitReady();
    } catch (error) { await this.stop(); throw error; }
  }
  private wgpuArgs(): string[] { return this.config.mode === "headless" ? ["--headless-server", this.config.wgpu] : ["--window-server", this.config.wgpu, this.config.ui, "--eventd", this.config.eventd]; }
  private async findExecutableDir(): Promise<string> {
    const requested = this.config.profile;
    const profiles: Array<Exclude<RuntimeProfile, "auto">> = requested === "auto" ? ["release", "debug"] : [requested];
    for (const profile of profiles) {
      const directory = join(this.config.neonRoot, "target", profile);
      try {
        await Promise.all(["neon-eventd.exe", "neon-wgpu-runtime.exe", "neon-ui-runtime.exe"].map((name) => access(join(directory, name))));
        this.selectedProfile = profile;
        return directory;
      } catch {
        // Try the next profile so a source checkout remains convenient in development.
      }
    }
    if ((requested === "auto" || requested === "release") && !process.env.NEON_ROOT && this.config.neonRoot === defaultNeonRoot(this.config.runtimeVersion)) {
      const downloadedRoot = await this.ensureDownloadedRuntime(this.config.runtimeVersion);
      this.selectedProfile = "release";
      return join(downloadedRoot, "target", "release");
    }
    throw new Error(`Neon3 runtime binaries not found under ${this.config.neonRoot}/target (${profiles.join(", ")})`);
  }
  private async ensureDownloadedRuntime(version: string): Promise<string> {
    const cacheRoot = join(process.env.LOCALAPPDATA ?? join(homedir(), "AppData", "Local"), "Neon3Sdk", "runtime", version);
    const binaries = ["neon-eventd.exe", "neon-wgpu-runtime.exe", "neon-ui-runtime.exe"];
    try { await Promise.all(binaries.map(name => access(join(cacheRoot, "target", "release", name)))); return cacheRoot; } catch { /* download below */ }
    await mkdir(cacheRoot, { recursive: true });
    const asset = NEON3_RUNTIME_ASSET(version);
    const archive = join(cacheRoot, `${asset}.download`);
    const url = `https://github.com/${NEON3_RUNTIME_REPOSITORY}/releases/download/${version}/${asset}`;
    const downloadTimeoutMs = Math.max(this.config.timeoutMs, 180000);
    await download(url, archive, downloadTimeoutMs);
    await new Promise<void>((resolve, reject) => execFile("tar", ["-xf", archive, "-C", cacheRoot], { timeout: downloadTimeoutMs }, error => error ? reject(error) : resolve()));
    await unlink(archive).catch(() => undefined);
    await Promise.all(binaries.map(name => access(join(cacheRoot, "target", "release", name))));
    return cacheRoot;
  }
  private async waitReady(): Promise<void> {
    const deadline = Date.now() + this.config.timeoutMs;
    for (const [target, endpoint] of [["eventd", this.config.eventd], ["wgpu-runtime", this.config.wgpu], ["ui-runtime", this.config.ui]] as const) {
      while (Date.now() < deadline) {
        try {
          if ((await new NeonClient(endpoint, { timeoutMs: 500 }).health(target)).status === "healthy") break;
        } catch {}
        await new Promise((resolve) => setTimeout(resolve, 100));
      }
      if (Date.now() >= deadline) throw new Error(`Timed out waiting for ${target} on ${endpoint}`);
    }
  }
  async stop(): Promise<void> { for (const process of this.processes.reverse()) { if (!process.killed) process.kill(); } this.processes = []; }
}

function defaultNeonRoot(version = process.env.NEON3_RUNTIME_VERSION ?? NEON3_RUNTIME_VERSION): string {
  if (process.env.NEON_ROOT) return process.env.NEON_ROOT;
  return join(process.env.LOCALAPPDATA ?? join(homedir(), "AppData", "Local"), "Neon3Sdk", "runtime", version);
}

async function download(url: string, destination: string, timeoutMs: number): Promise<void> {
  const proxy = process.env.HTTPS_PROXY ?? process.env.HTTP_PROXY ?? process.env.https_proxy ?? process.env.http_proxy;
  const response = await fetch(url, {
    signal: AbortSignal.timeout(timeoutMs),
    redirect: "follow",
    ...(proxy ? { dispatcher: new ProxyAgent(proxy) } : {}),
  } as RequestInit);
  if (!response.ok) throw new Error(`Neon3 runtime download failed: HTTP ${response.status}`);
  await writeFile(destination, Buffer.from(await response.arrayBuffer()));
}

async function resolveLatestRuntimeVersion(): Promise<string> {
  const proxy = process.env.HTTPS_PROXY ?? process.env.HTTP_PROXY ?? process.env.https_proxy ?? process.env.http_proxy;
  const options = { signal: AbortSignal.timeout(30000), ...(proxy ? { dispatcher: new ProxyAgent(proxy) } : {}) } as RequestInit;
  const response = await fetch(`https://api.github.com/repos/${NEON3_RUNTIME_REPOSITORY}/releases/latest`, { ...options, headers: { Accept: "application/vnd.github+json", "User-Agent": "@neon3/sdk" } } as RequestInit);
  if (response.ok) {
    const release = await response.json() as { tag_name?: unknown };
    if (typeof release.tag_name === "string" && release.tag_name.length > 0) return release.tag_name;
  }
  const page = await fetch(`https://github.com/${NEON3_RUNTIME_REPOSITORY}/releases/latest`, { ...options, redirect: "manual" } as RequestInit);
  const location = page.headers.get("location") ?? "";
  const marker = "/releases/tag/";
  const tag = location.includes(marker) ? location.split(marker, 2)[1].split("/", 1)[0] : "";
  if (!tag) throw new Error(`Unable to resolve latest Neon3 runtime release: API HTTP ${response.status}`);
  return tag;
}

File: packages/python-sdk/src/neon3_sdk/runtime.py
"""Runtime mode and process configuration types for library users."""

from __future__ import annotations

import enum
import json
import os
import subprocess
import time
import urllib.request
import zipfile
from pathlib import Path
from dataclasses import dataclass, field
from typing import Any

from .client import NeonClient

NEON3_RUNTIME_VERSION = "latest"
NEON3_RUNTIME_REPOSITORY = "unco999/Neon3-CiJian"
NEON3_RUNTIME_ASSET_TEMPLATE = "neon3-runtime-windows-x86_64-{version}.zip"


def runtime_version() -> str:
    """Return the selected runtime release, defaulting to GitHub latest."""
    return os.environ.get("NEON3_RUNTIME_VERSION", NEON3_RUNTIME_VERSION)


def _resolve_runtime_version(version: str) -> str:
    if version != "latest":
        return version
    request = urllib.request.Request(
        f"https://api.github.com/repos/{NEON3_RUNTIME_REPOSITORY}/releases/latest",
        headers={"Accept": "application/vnd.github+json", "User-Agent": "neon3-sdk"},
    )
    with urllib.request.urlopen(request, timeout=30) as response:
        release = json.loads(response.read().decode("utf-8"))
    tag = release.get("tag_name")
    if not isinstance(tag, str) or not tag.strip():
        raise RuntimeError("GitHub latest Neon3 release did not contain tag_name")
    return tag


def default_neon_root(version: str | None = None) -> Path:
    """Return an explicit root, local SDK bundle, or per-user runtime cache."""
    override = os.environ.get("NEON_ROOT")
    if override:
        return Path(override)
    local_app_data = Path(os.environ.get("LOCALAPPDATA", Path.home() / "AppData" / "Local"))
    return local_app_data / "Neon3Sdk" / "runtime" / (version or runtime_version())


def _runtime_available(root: Path) -> bool:
    return all((root / "target" / "release" / name).is_file() for name in (
        "neon-eventd.exe", "neon-wgpu-runtime.exe", "neon-ui-runtime.exe"
    ))


def _download_runtime(root: Path, version: str) -> None:
    root.mkdir(parents=True, exist_ok=True)
    asset = NEON3_RUNTIME_ASSET_TEMPLATE.format(version=version)
    archive = root.parent / f"{asset}.download"
    url = f"https://github.com/{NEON3_RUNTIME_REPOSITORY}/releases/download/{version}/{asset}"
    try:
        with urllib.request.urlopen(url, timeout=180) as response, archive.open("wb") as stream:
            stream.write(response.read())
        with zipfile.ZipFile(archive) as bundle:
            bundle.extractall(root)
    finally:
        archive.unlink(missing_ok=True)


class RuntimeMode(str, enum.Enum):
    WINDOWED = "windowed"
    HEADLESS = "headless"
    EXTERNAL_SURFACE = "external_surface"


@dataclass(frozen=True)
class RuntimeEndpoints:
    eventd: str = "127.0.0.1:39101"
    ui: str = "127.0.0.1:39102"
    wgpu: str = "127.0.0.1:39103"


@dataclass(frozen=True)
class RuntimeConfig:
    neon_root: str = field(default_factory=lambda: str(default_neon_root()))
    mode: RuntimeMode = RuntimeMode.WINDOWED
    endpoints: RuntimeEndpoints = RuntimeEndpoints()
    domain_endpoint: str = "127.0.0.1:39104"
    timeout_seconds: float = 15.0
    profile: str = field(default_factory=lambda: os.environ.get("NEON_PROFILE", "auto"))
    runtime_version: str = field(default_factory=runtime_version)

    @property
    def wgpu_arguments(self) -> tuple[str, ...]:
        if self.mode is RuntimeMode.WINDOWED:
            return ("--window-server", self.endpoints.wgpu, self.endpoints.ui, "--eventd", self.endpoints.eventd)
        if self.mode is RuntimeMode.EXTERNAL_SURFACE:
            return ("--window-server", self.endpoints.wgpu, self.endpoints.ui, "--eventd", self.endpoints.eventd)
        return ("--headless-server", self.endpoints.wgpu)


class RuntimeSession:
    """Owns only processes started by the SDK and exposes deterministic lifecycle."""

    def __init__(self, config: RuntimeConfig) -> None:
        self.config = config
        self.processes: list[tuple[str, subprocess.Popen[str]]] = []

    def start(self) -> None:
        selected_runtime_version = self.config.runtime_version
        if selected_runtime_version == "latest":
            selected_runtime_version = _resolve_runtime_version(selected_runtime_version)
        root = Path(self.config.neon_root)
        if not os.environ.get("NEON_ROOT") and self.config.neon_root == str(default_neon_root("latest")):
            root = default_neon_root(selected_runtime_version)
        requested_profile = self.config.profile.lower()
        if requested_profile not in {"auto", "release", "debug"}:
            raise ValueError("profile must be auto, release, or debug")
        profiles = (requested_profile,) if requested_profile != "auto" else ("release", "debug")
        runtime_dir = next(
            (
                root / "target" / profile
                for profile in profiles
                if all((root / "target" / profile / name).is_file() for name in ("neon-eventd.exe", "neon-wgpu-runtime.exe", "neon-ui-runtime.exe"))
            ),
            None,
        )
        can_download = (
            runtime_dir is None
            and self.config.profile.lower() in {"auto", "release"}
            and not os.environ.get("NEON_ROOT")
            and root == default_neon_root(selected_runtime_version)
        )
        if can_download:
            _download_runtime(root, selected_runtime_version)
            runtime_dir = root / "target" / "release"
        if runtime_dir is None:
            raise FileNotFoundError(f"Neon3 release/debug binaries not found under {root}")
        specs = [
            ("eventd", runtime_dir / "neon-eventd.exe", ("--server", self.config.endpoints.eventd, "1")),
            ("wgpu-runtime", runtime_dir / "neon-wgpu-runtime.exe", self.config.wgpu_arguments),
            ("ui-runtime", runtime_dir / "neon-ui-runtime.exe", ("--forward-server", self.config.endpoints.ui, self.config.endpoints.wgpu, self.config.domain_endpoint, "--eventd", self.config.endpoints.eventd)),
        ]
        try:
            for name, executable, arguments in specs:
                if not executable.is_file():
                    raise FileNotFoundError(f"Neon3 executable not found: {executable}")
                process = subprocess.Popen([str(executable), *arguments], cwd=root, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, text=True)
                self.processes.append((name, process))
            self.wait_ready()
        except Exception:
            self.stop()
            raise

    def wait_ready(self) -> None:
        deadline = time.monotonic() + self.config.timeout_seconds
        for target, endpoint in (("eventd", self.config.endpoints.eventd), ("wgpu-runtime", self.config.endpoints.wgpu), ("ui-runtime", self.config.endpoints.ui)):
            while time.monotonic() < deadline:
                try:
                    health = NeonClient.connect(endpoint, origin="neon3-sdk-runtime", timeout_seconds=0.5).health(target)
                    if health.status == "healthy":
                        break
                except Exception:
                    pass
                time.sleep(0.1)
            else:
                raise TimeoutError(f"timed out waiting for {target} at {endpoint}")

    def stop(self) -> None:
        for _name, process in reversed(self.processes):
            if process.poll() is None:
                process.terminate()
                try:
                    process.wait(timeout=3)
                except subprocess.TimeoutExpired:
                    process.kill()
                    process.wait(timeout=3)
        self.processes.clear()

    def __enter__(self) -> "RuntimeSession":
        self.start()
        return self

    def __exit__(self, _type: Any, _value: Any, _traceback: Any) -> None:
        self.stop()
